### PR TITLE
feat: compute a better gas limit for recursive external contract calls 

### PR DESCRIPTION
**Pull request by travisperson**
## Related Issues

https://github.com/filecoin-project/lotus/issues/10041

## Proposed Changes

After running `GasEstimateMessageGas` in `EthEstimateGas` execute the message using `CallWithGas` to verify that the estimated gas limit (derived from the gas used) succeeds. If the message fails, double the gas limit till success is achieved (usually within 1), and then binary search to within 100,000 gas, always returning the last successful gas limit that succeeded.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: node/impl/full/eth.go [hunk 1]
```diff
@@ -811,12 +812,161 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
...
+		low = high
+		high = high * 2
+
+		if high > build.BlockGasLimit {
+			return -1, xerrors.Errorf("No message successful")
+		}
+	}
+
+	log.Infow("findGoodGas narrow down to best GasLimit")
+	for (high - low) > 100_000 {
```
Comment (travisperson) [inline node/impl/full/eth.go:884]: Is there a better way to find a "close enough"?

File: node/impl/full/eth.go [hunk 2]
```diff
@@ -811,12 +812,161 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
...
+		if ok {
+			break
+		}
+
+		low = high
+		high = high * 2
+
+		if high > build.BlockGasLimit {
+			return -1, xerrors.Errorf("No message successful")
+		}
```
Comment (travisperson) [inline node/impl/full/eth.go:880]: Set to build.BlockGasLimit and break

File: node/impl/full/eth.go [hunk 3]
```diff
@@ -811,12 +812,168 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
...
+			break
+		}
+		ts, err = cstore.GetTipSetFromKey(ctx, ts.Parents())
+		if err != nil {
+			return -1, xerrors.Errorf("getting parent tipset: %w", err)
+		}
+	}
+	if err != nil {
+		return -1, xerrors.Errorf("CallWithGas failed: %w", err)
+	}
```
Comment (travisperson) [inline node/impl/full/eth.go:957]: From 929 down to here is a copy of https://github.com/filecoin-project/lotus/blob/d6da796bf2d36560ad5b0a49a871d8e121ca8df4/node/impl/full/gas.go#L264-L292

I was thinking of pulling this out into its own function, but I would also need to return `priorMsgs` to be passed down further, as I need them when making additional calls to `CallWithSGas` inside of the currently named function `findGoodGas` which does the search.

So my time boxes refactor is to split the following into their own functions and use them here, and inside of `gasEstimateGasLimit`.

returns `priorMsgs`
https://github.com/filecoin-project/lotus/blob/d6da796bf2d36560ad5b0a49a871d8e121ca8df4/node/impl/full/gas.go#L269-L276

returns `res, err`
https://github.com/filecoin-project/lotus/blob/d6da796bf2d36560ad5b0a49a871d8e121ca8df4/node/impl/full/gas.go#L278-L292

Comment (travisperson) [reply node/impl/full/eth.go:937]: Actually there is probably not much of a reason to have two methods, and probably should just extract that entire block.

File: node/impl/full/eth.go [hunk 4]
```diff
@@ -804,12 +805,137 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
 	// gas estimation actually run.
 	msg.GasLimit = 0
 
-	msg, err = a.GasAPI.GasEstimateMessageGas(ctx, msg, nil, types.EmptyTSK)
+	ts := a.Chain.GetHeaviestTipSet()
+	msg, err = a.GasAPI.GasEstimateMessageGas(ctx, msg, nil, ts.Key())
 	if err != nil {
 		return ethtypes.EthUint64(0), err
```
Comment (arajasek) [inline node/impl/full/eth.go:811]: Here, and in other places, can we please chain errors? It makes debugging much easier.
```suggestion
		return ethtypes.EthUint64(0), xerrors.Errorf("failed to estimate gas: %w", err)
```

File: node/impl/full/eth.go [hunk 5]
```diff
@@ -804,12 +805,137 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
...
+) (int64, error) {
+	msg := *msgIn
+	currTs := ts
+
+	res, priorMsgs, ts, err := gasEstimateCallWithGas(ctx, cstore, smgr, mpool, &msg, currTs)
+	if err != nil {
+		return -1, err
+	}
+
+	if res.MsgRct.ExitCode == exitcode.Ok {
```
Comment (arajasek) [inline node/impl/full/eth.go:924]:
```suggestion
	if res.MsgRct.ExitCode.IsSuccess() {
```

File: node/impl/full/eth.go [hunk 6]
```diff
@@ -804,12 +805,137 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
...
+	mpool *messagepool.MessagePool,
+	msgIn *types.Message,
+	ts *types.TipSet,
+) (int64, error) {
+	msg := *msgIn
+	currTs := ts
+
+	res, priorMsgs, ts, err := gasEstimateCallWithGas(ctx, cstore, smgr, mpool, &msg, currTs)
+	if err != nil {
+		return -1, err
```
Comment (arajasek) [inline node/impl/full/eth.go:921]: chain error

File: node/impl/full/eth.go [hunk 7]
```diff
@@ -804,12 +805,137 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
...
+	}
+
+	if res.MsgRct.ExitCode == exitcode.Ok {
+		return msg.GasLimit, nil
+	}
+
+	if traceContainsExitCode(res.ExecutionTrace, exitcode.SysErrOutOfGas) {
+		ret, err := gasSearch(ctx, smgr, &msg, priorMsgs, ts)
+		if err != nil {
+			return -1, err
```
Comment (arajasek) [inline node/impl/full/eth.go:931]: chain error

File: node/impl/full/eth.go [hunk 8]
```diff
@@ -804,12 +805,137 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
...
+	if err != nil {
+		log.Errorw("expected gas", "err", err)
+	}
+
+	return ethtypes.EthUint64(expectedGas), nil
+}
+
+// gasSearch does an exponential search to find a gas value to execute the
+// message with. It first finds a high gas limit that allows the message to execute
+// by doubling the previous gas limit until it successeds then does a binary
```
Comment (arajasek) [inline node/impl/full/eth.go:824]:
```suggestion
// by doubling the previous gas limit until it succeeds, then does a binary
```

File: node/impl/full/eth.go [hunk 9]
```diff
@@ -804,12 +805,137 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
...
+
+	canSuccessed := func(limit int64) (bool, error) {
+		msg.GasLimit = limit
+
+		res, err := smgr.CallWithGas(ctx, &msg, priorMsgs, ts)
+		if err != nil {
+			return false, err
+		}
+
+		if res.MsgRct.ExitCode == exitcode.Ok {
```
Comment (arajasek) [inline node/impl/full/eth.go:846]:
```suggestion
		if res.MsgRct.ExitCode.IsSuccess() {
```

File: node/impl/full/eth.go [hunk 10]
```diff
@@ -804,12 +805,137 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
...
+		low = high
+		high = high * 2
+
+		if high > build.BlockGasLimit {
+			high = build.BlockGasLimit
+			break
+		}
+	}
+
+	checkThreshold := (high / 100)
```
Comment (arajasek) [inline node/impl/full/eth.go:871]: Drop the parens?

File: node/impl/full/eth.go [hunk 11]
```diff
@@ -804,12 +805,137 @@ func (a *EthModule) EthEstimateGas(ctx context.Context, tx ethtypes.EthCall) (et
...
+	msgIn *types.Message,
+	priorMsgs []types.ChainMsg,
+	ts *types.TipSet,
+) (int64, error) {
+	msg := *msgIn
+
+	high := msg.GasLimit
+	low := msg.GasLimit
+
+	canSuccessed := func(limit int64) (bool, error) {
```
Comment (arajasek) [inline node/impl/full/eth.go:838]: `canSucceed`?

File: node/impl/full/gas.go [hunk 12]
```diff
@@ -248,22 +248,19 @@ func (m *GasModule) GasEstimateGasLimit(ctx context.Context, msgIn *types.Messag
 	}
 	return gasEstimateGasLimit(ctx, m.Chain, m.Stmgr, m.Mpool, msgIn, ts)
 }
-func gasEstimateGasLimit(
+
+func gasEstimateCallWithGas(
```
Comment (arajasek) [inline node/impl/full/gas.go:252]: Can we add a comment describing the return here (specifically the `priorMsgs` bit).

**Review by arajasek — COMMENTED** [review]
Basically good to go, just some nits.

**Review by arajasek — APPROVED** [review]

**Pull request merged**